### PR TITLE
commands_cache: ignore mtime check on non-darwin os

### PR DESCRIPTION
**Pull request by yotamolenik**
darwin's stat uses nano seconds accuracy for the files mtime property, whereas
on linux it returns a unix timestamp (seconds), meaning we cannot rely on that
field when checking for dir modifications per each path for creating `CommandsCache`.

consider the following shell script:

```shell
cp /path/to/executable /bin/new_executable

new_executable
```

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

### CODE REVIEW

**Comment by jnoortheen** [comment]
Thanks @yotamolenik for trying to fix this. I was working on `CommandsCache` and this fix was part of #4954 . Now it is ready, could you test that PR?

**Comment by yotamolenik** [comment]
@jnoortheen the bug persists on linux, but I rebased my changes on top of your branch and tested the fix.
it works now on both linux and darwin 🥳

**Comment by jnoortheen** [comment]
@yotamolenik Ok thats good. Lets wait until #4954 is merged.

File: xonsh/commands_cache.py
Comment (jnoortheen) [inline xonsh/commands_cache.py:169]: the xonsh.platform has already similar functions. Lets reuse that.

Comment (yotamolenik) [reply xonsh/commands_cache.py]: done

**Comment by jnoortheen** [comment]
Not using cache at all seems a bit too much. It may lead to shell being sluggish. Perhaps we can add a warning in the doc?

**Comment by doronz88** [comment]
@jnoortheen what do you propose instead of not using cache on non-darwin OS? i think sluggish is better than a shell that relies on something with chances of failure.
the worst thing about this bug is that it may resolve in really bad errors. i encountered this bug when i had a python entry point both inside a venv and on global path.
it sometimes resolved the entry point to the system-one and sometimes to the venv one. only workarround is adding a `sleep 1` after the venv activation

**Comment by jnoortheen** [comment]
I think we can use a setting value to control this behaviour while defaulting to being enabled. We can show the warning. 
Previously there was a bug which caused stale results. It was fixed with  #4954 .

**Comment by jnoortheen** [comment]
The branch is now merged. You can start with `xonsh/environ.py`. I am on mobile, just pointing out the file name.

**Comment by yotamolenik** [comment]
@jnoortheen Do you mean like this?

File: xonsh/environ.py
Comment (doronz88) [inline xonsh/environ.py:871]: reformat. it ensures executables' path will be resolved to correct one on platforms where the directories `mtime` field isn't represented accurate enough (nanoseconds on Darwin, but seconds on Linux).

**Review by doronz88 — CHANGES_REQUESTED** [review]

File: xonsh/commands_cache.py
Comment (jnoortheen) [inline xonsh/commands_cache.py:174]: this should be , `self.env.get('COMMANDS_CACHE_IGNORE_MTIME_CHECK')` .

File: xonsh/environ.py
Comment (jnoortheen) [inline xonsh/environ.py:870]: - the default can be False. 
- Ultimately this variable disables commands_cache's cache system not just the mtime check. mtime-check is an implementation detail. I think this should be named `ENABLE_COMMANDS_CACHE(default=True)` . I suggest the description to be `command names in a directory are cached when enabled. It makes use of directory's modified time. On some platforms it may not be accurate enough (e.g. Windows, Linux save mtime in seconds). setting it to False would disable the caching mechanism and may slow down the shell`

Comment (yotamolenik) [reply xonsh/environ.py]: sounds good!
I also changed my test a little bit to run only if the cache is disabled.

File: tests/test_commands_cache.py
Comment (jnoortheen) [inline tests/test_commands_cache.py:180]: the test should set the variable to False and proceed. You can see the other tests in this file for similar tests.

File: xonsh/commands_cache.py
Comment (jnoortheen) [inline xonsh/commands_cache.py:172]: it is best to pass the default value True here to `.get()`

File: tests/test_commands_cache.py
Comment (yotamolenik) [reply tests/test_commands_cache.py]: done

File: xonsh/commands_cache.py
Comment (yotamolenik) [reply xonsh/commands_cache.py:172]: Why is it best?🤔
If the value doesn't exist it should be a bug

Comment (jnoortheen) [reply xonsh/commands_cache.py:172]: sometimes the variable might be None; during unit tests.

File: xonsh/commands_cache.py
Comment (doronz88) [inline xonsh/commands_cache.py:172]: the condition is the exact opposite. currently if its enabled, it means to disable

**Review by doronz88 — CHANGES_REQUESTED** [review]

**Review by doronz88 — APPROVED** [review]

File: tests/test_commands_cache.py
Comment (doronz88) [inline tests/test_commands_cache.py:180]: redundant

File: xonsh/commands_cache.py
Comment (yotamolenik) [reply xonsh/commands_cache.py]: done

**Comment by yotamolenik** [comment]
I believe the code is ok now. 
About the linux failure it might be related to the os kernel and the scheduler for the time.sleep() function
About the windows failure I made a tiny mistake with the name of the exe - fixed it now and it should work

**Comment by jnoortheen** [comment]
@yotamolenik Please make sure that the test passes on Linux as well.

**Comment by yotamolenik** [comment]
It should always pass now, i did have another mistake

**Comment by jnoortheen** [comment]
pre-commit.ci autofix

File: xonsh/environ.py
Comment (jnoortheen) [inline xonsh/environ.py:873]: it is good to define `doc_default` as well. It will appear in the docs generated

Comment (yotamolenik) [reply xonsh/environ.py:873]: do you mean like this?

Comment (jnoortheen) [reply xonsh/environ.py:873]: I've updated it.

**Review by jnoortheen — APPROVED** [review]

**Comment by jnoortheen** [comment]
Thanks @yotamolenik for the continued effort!

**Pull request merged**
